packages/block-editor/src/components/block-navigation/use-block-navigation-client-ids.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * WordPress dependencies
+ */
+
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { isClientIdSelected } from './utils';
+import { store as blockEditorStore } from '../../store';
+
+const useBlockNavigationSelectedClientIds = (
+	selectedBlockClientIds,
+	__experimentalPersistentListViewFeatures
+) =>
+	useSelect(
+		( select ) => {
+			const {
+				getSelectedBlockClientId,
+				getSelectedBlockClientIds,
+			} = select( blockEditorStore );
+
+			if ( selectedBlockClientIds ) {
+				return selectedBlockClientIds;
+			}
+
+			if ( __experimentalPersistentListViewFeatures ) {
+				return getSelectedBlockClientIds();
+			}
+
+			return getSelectedBlockClientId();
+		},
+		[ selectedBlockClientIds, __experimentalPersistentListViewFeatures ]
+	);
+
+const useBlockNavigationClientIdsTree = (
+	blocks,
+	selectedClientIds,
+	showOnlyCurrentHierarchy
+) =>
+	useSelect(
+		( select ) => {
+			const {
+				getBlockHierarchyRootClientId,
+				__unstableGetClientIdsTree,
+				__unstableGetClientIdWithClientIdsTree,
+			} = select( blockEditorStore );
+
+			if ( blocks ) {
+				return blocks;
+			}
+
+			const isSingleBlockSelected =
+				selectedClientIds && ! Array.isArray( selectedClientIds );
+			if ( ! showOnlyCurrentHierarchy || ! isSingleBlockSelected ) {
+				return __unstableGetClientIdsTree();
+			}
+
+			const rootBlock = __unstableGetClientIdWithClientIdsTree(
+				getBlockHierarchyRootClientId( selectedClientIds )
+			);
+			if ( ! rootBlock ) {
+				return __unstableGetClientIdsTree();
+			}
```

Comment:
I'd avoid that. If we merge, then we are calling the `__unstableGetClientIdWithClientIdsTree` selector all the time. Rather than calling it when we just need it.

Reply:
Yeah, as @david-szabo97 said, it's a simple optimization.
I've tried for a while to come up with a more elegant solution, but I wasn't able to find one.